### PR TITLE
Impose positive stddev in Gaussian models for fitting

### LINKED ISSUE #1105: Default Gaussian fitting should not allow negative sigma
I tried out `astropy.modeling` for the first time tonight, and it works great! However, I was puzzled to see that half the time my Gaussian fits end up with a negative sigma, which is unexpected, so I was wondering whether the Gaussian model could have a default constraint on `sigma` that it should be positive?

### LINKED ISSUE #6014: Negative standard deviation from Gaussian fit
It has already been pointed out in #1105.
I have the following 17 by 17 array called `test`, and I did a Gaussian 2D fitting and got negative `x_stddev`:

```python
test = [[-54.33, 13.81, -34.55, 8.95, -143.71, -0.81, 59.25, -14.78, -204.9, -30.87, -124.39, 123.53, 70.81, -109.48, -106.77, 35.64, 18.29], [-126.19, -89.13, 63.13, 50.74, 61.83, 19.06, 65.7, 77.94, 117.14, 139.37, 52.57, 236.04, 100.56, 242.28, -180.62, 154.02, -8.03], [91.43, 96.45, -118.59, -174.58, -116.49, 80.11, -86.81, 14.62, 79.26, 7.56, 54.99, 260.13, -136.42, -20.77, -77.55, 174.52, 134.41], [33.88, 7.63, 43.54, 70.99, 69.87, 33.97, 273.75, 176.66, 201.94, 336.34, 340.54, 163.77, -156.22, 21.49, -148.41, 94.88, 42.55], [82.28, 177.67, 26.81, 17.66, 47.81, -31.18, 353.23, 589.11, 553.27, 242.35, 444.12, 186.02, 140.73, 75.2, -87.98, -18.23, 166.74], [113.09, -37.01, 134.23, 71.89, 107.88, 198.69, 273.88, 626.63, 551.8, 547.61, 580.35, 337.8, 139.8, 157.64, -1.67, -26.99, 37.35], [106.47, 31.97, 84.99, -125.79, 195.0, 493.65, 861.89, 908.31, 803.9, 781.01, 532.59, 404.67, 115.18, 111.11, 28.08, 122.05, -58.36], [183.62, 45.22, 40.89, 111.58, 425.81, 321.53, 545.09, 866.02, 784.78, 731.35, 609.01, 405.41, -19.65, 71.2, -140.5, 144.07, 25.24], [137.13, -86.95, 15.39, 180.14, 353.23, 699.01, 1033.8, 1014.49, 814.11, 647.68, 461.03, 249.76, 94.8, 41.17, -1.16, 183.76, 188.19], [35.39, 26.92, 198.53, -37.78, 638.93, 624.41, 816.04, 867.28, 697.0, 491.56, 378.21, -18.46, -65.76, 98.1, 12.41, -102.18, 119.05], [190.73, 125.82, 311.45, 369.34, 554.39, 454.37, 755.7, 736.61, 542.43, 188.24, 214.86, 217.91, 7.91, 27.46, -172.14, -82.36, -80.31], [-55.39, 80.18, 267.19, 274.2, 169.53, 327.04, 488.15, 437.53, 225.38, 220.94, 4.01, -92.07, 39.68, 57.22, 144.66, 100.06, 34.96], [130.47, -4.23, 46.3, 101.49, 115.01, 217.38, 249.83, 115.9, 87.36, 105.81, -47.86, -9.94, -82.28, 144.45, 83.44, 23.49, 183.9], [-110.38, -115.98, 245.46, 103.51, 255.43, 163.47, 56.52, 33.82, -33.26, -111.29, 88.08, 193.2, -100.68, 15.44, 86.32, -26.44, -194.1], [109.36, 96.01, -124.89, -16.4, 84.37, 114.87, -65.65, -58.52, -23.22, 42.61, 144.91, -209.84, 110.29, 66.37, -117.85, -147.73, -122.51], [10.94, 45.98, 118.12, -46.53, -72.14, -74.22, 21.22, 0.39, 86.03, 23.97, -45.42, 12.05, -168.61, 27.79, 61.81, 84.07, 28.79], [46.61, -104.11, 56.71, -90.85, -16.51, -66.45, -141.34, 0.96, 58.08, 285.29, -61.41, -9.01, -323.38, 58.35, 80.14, -101.22, 145.65]]
g_init = Gaussian2D(x_mean=8, y_mean=8)
fitter = LevMarLSQFitter()
y, x = np.mgrid[:17,:17]
g_fit = fitter(g_init, x, y, test)
#  <Gaussian2D(
#amplitude=984.7694929790363, 
#x_mean=7.198391516587464, 
#y_mean=7.49720660088511,
#x_stddev=-1.9840185107597297, 
#y_stddev=3.1840618351417307, 
#theta=-18.144136589252792)>
```

So here I have two questions:
* Can I just take absolute value of it?
* Why do `theta` has so large absolute value, not constrained in (-pi, +pi)?  
  * I didn't show it here, but if I constrain `theta` to some arbitrary value, such as (-tt1, +tt2), I get only -tt1 or +tt2, i.e., the boundary values. It seems like the fitter could not find the local minimum well, unless unconstrained. Is there any reason for this?

### PR DESCRIPTION
**Pull request by pllim**
Fix #6014 
Fix #1105

Supersede #1574

c/c @nden @astrofrog @cdeil @ysBach

Note: See https://travis-ci.org/pllim/astropy/builds/242949773 for Travis results. I canceled Travis on purpose due to long queue.

### CODE REVIEW

**Comment by pllim** [comment]
This failed on Appveyor:
```
________ [doctest] astropy.stats.info_theory.akaike_info_criterion_lsq ________
372     ...            models.Gaussian1D(2, 0.5, 0.1))
373     >>> g2_fit = fitter(g2_init, x, y)
374     >>> # Fit with only one Gaussian
375     >>> g1_init = models.Gaussian1D(amplitude=2., mean=0.3, stddev=.5)
376     >>> g1_fit = fitter(g1_init, x, y)
377     >>> # Compute the mean squared errors
378     >>> ssr_g3 = np.sum((g3_fit(x) - y)**2.0)
379     >>> ssr_g2 = np.sum((g2_fit(x) - y)**2.0)
380     >>> ssr_g1 = np.sum((g1_fit(x) - y)**2.0)
381     >>> akaike_info_criterion_lsq(ssr_g3, 9, x.shape[0]) # doctest: +FLOAT_CMP
Expected:
    -656.32589850659224
Got:
    -660.41075962618652
astropy\stats\info_theory.py:381: DocTestFailure
 1 failed, 11655 passed, 204 skipped, 81 xfailed, 1 xpassed in 296.44 seconds =
Command exited with code 1
```

**Comment by cdeil** [comment]
I'm swamped with other stuff, unsubscribing now. Just wanted to say +1.

File: astropy/modeling/tests/test_constraints.py [hunk 1]
```diff
@@ -424,3 +424,56 @@ def evaluate(x, a, b):
...
+    g_init = models.Gaussian2D(x_mean=8, y_mean=8)
+    fitter = fitting.LevMarLSQFitter()
+    y, x = np.mgrid[:17, :17]
+    g_fit = fitter(g_init, x, y, test)
+
+    # Compare with @ysBach original result:
+    # - x_stddev was negative, so its abs value is used for comparison here.
+    # - theta is beyond (-90, 90) deg, which doesn't make sense, so ignored.
+    utils.assert_allclose([g_fit.amplitude.value, g_fit.y_stddev.value],
+                          [984.7694929790363, 3.1840618351417307], rtol=1.5e-6)
```
Comment (pllim) [inline astropy/modeling/tests/test_constraints.py]: @ysBach are these "rtol" values acceptable?

File: astropy/stats/info_theory.py [hunk 2]
```diff
@@ -379,7 +379,7 @@ def akaike_info_criterion_lsq(ssr, n_params, n_samples):
     >>> ssr_g2 = np.sum((g2_fit(x) - y)**2.0)
     >>> ssr_g1 = np.sum((g1_fit(x) - y)**2.0)
     >>> akaike_info_criterion_lsq(ssr_g3, 9, x.shape[0]) # doctest: +FLOAT_CMP
-    -656.32589850659224
+    -660.41075962620482
```
Comment (pllim) [inline astropy/stats/info_theory.py]: @mirca , this result has changed and history shows you added this. Is this change acceptable?

File: astropy/modeling/tests/test_constraints.py (hunk 1 above)
Comment (ysBach) [inline astropy/modeling/tests/test_constraints.py]: I think yes :)

File: astropy/modeling/functional_models.py [hunk 3]
```diff
@@ -33,6 +35,17 @@ class BaseGaussian1D(Fittable1DModel):
     mean = Parameter(default=0)
     stddev = Parameter(default=1)
 
+    def __init__(self, amplitude, mean, stddev, **kwargs):
+        # Ensure stddev makes sense if its bounds are not explicitly set.
+        if 'bounds' not in kwargs:
+            kwargs['bounds'] = {'stddev': (FLOAT_EPSILON, None)}
```
Comment (astrofrog) [inline astropy/modeling/functional_models.py]: You should add a comment here and for the 2D Gaussian to explain why not simply ``(0, None)``, just for anyone looking the code in future.

File: astropy/modeling/functional_models.py [hunk 4]
```diff
@@ -33,6 +35,17 @@ class BaseGaussian1D(Fittable1DModel):
     mean = Parameter(default=0)
     stddev = Parameter(default=1)
```
Comment (astrofrog) [inline astropy/modeling/functional_models.py]: Just out of curiosity, why not simply set ``Parameter(default=1, bounds=(FLOAT_EPSILON, None))`` here - is it because the user then can't override the bounds?

File: astropy/modeling/tests/test_constraints.py [hunk 5]
```diff
@@ -424,3 +424,56 @@ def evaluate(x, a, b):
     # Check that the estimated Jacobian was computed (it doesn't matter what
     # the values are so long as they're not all zero.
     assert np.any(f2.fit_info['fjac'] != 0)
+
+
+# https://github.com/astropy/astropy/issues/6014
+@pytest.mark.skipif('not HAS_SCIPY')
+def test_gaussian2d_positive_stddev():
+    test = [
```
Comment (astrofrog) [inline astropy/modeling/tests/test_constraints.py]: Should include a comment to explain what this data is

**Comment by nden** [comment]
@pllim Currently the sed in np.random is set as `rsn = RandomState(1234567890)` within the `TestNonLinearConstraints` class. Could you move this statement to the module level sio that it can be used by your test as well. The random seeding may account for the differences in values you and @ysBach get.

File: astropy/modeling/functional_models.py (hunk 3 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: Is "because zero stddev is non-physical" sufficient? If not, please provide some text...

File: astropy/modeling/functional_models.py (hunk 4 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: Maybe @nden can confirm. I was just adapting from incantations I found from discussions in other issue and PR.

File: astropy/modeling/functional_models.py (hunk 3 above)
Comment (astrofrog) [inline astropy/modeling/functional_models.py]: I wonder if we should have a more global way of specifying this constant then and have one recommended way for users to say that they want to specify the parameter should be >0

Comment (nden) [inline astropy/modeling/functional_models.py]: Just want to caution that in general depending on the model and how the parameter is used this may lead to overflow errors. In #5656 we ended up making it clear in the documentation that the parameter should be positive.

File: astropy/modeling/functional_models.py (hunk 4 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: > What happens when a user sets the bounds?

Then user takes full responsibility of weird results.

File: astropy/modeling/functional_models.py (hunk 3 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: > I wonder if we should have a more global way

That is out of scope of this PR. But we can move FLOAT_EPSILON to `utils` or something if you want.

**Comment by pllim** [comment]
> The random seeding may account for the differences

@nden, why would `rsn` make a difference in my case? I am not use any randomness in my data. Are you saying the fitting somehow uses a random seed?

File: astropy/modeling/tests/test_constraints.py (hunk 5 above)
Comment (pllim) [inline astropy/modeling/tests/test_constraints.py]: Done.

**Comment by nden** [comment]
@pllim Yes, sorry. I may have confused it with a different PR.

**Comment by pllim** [comment]
@nden I moved the random seed to module level anyway, as I think it is still a good idea to make it global for the module, for future tests.

**Comment by pllim** [comment]
I have no idea why CircleCI is failing. It is complaining about convolution `nan` vs zeroes, which I have not touched. I blame @larrybradley since he touched it with some Docker thingy.

**Comment by pllim** [comment]
Hmm. Nevermind. I can reproduce the convolution failures locally but I don't know why it is happening. Example...

```
astropy/convolution/kernels.py ...
astropy/convolution/tests/test_convolve.py ...............................................................................................................
astropy/convolution/tests/test_convolve_fft.py ...................................................................................................................................................................................................................................................................................................................................................
astropy/convolution/tests/test_convolve_kernels.py F...F...F...F...F...F...F...F...F...F...F...F...F...F...F...F.................
astropy/convolution/tests/test_convolve_nddata.py FFF
astropy/convolution/tests/test_discretize.py ...s........FFFs........
```
(and then it takes so long that I ran out of patience and killed the test)

File: astropy/modeling/functional_models.py (hunk 4 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: Setting bounds within `Parameter` does not work for 2D case. FYI.

**Comment by pllim** [comment]
@astrofrog @nden , please re-review the changes. I can also squash the commits once approved, if desired. All the tests now have passed locally.

File: astropy/modeling/functional_models.py [hunk 6]
```diff
@@ -349,6 +354,17 @@ def __init__(self, amplitude=amplitude.default, x_mean=x_mean.default,
...
+        # stddev must be non-zero and positive.
+        if 'bounds' not in kwargs:
+            kwargs['bounds'] = {'x_stddev': (FLOAT_EPSILON, None),
+                                'y_stddev': (FLOAT_EPSILON, None)}
+        else:
+            kwargs['bounds']['x_stddev'] = kwargs['bounds'].get(
+                'x_stddev', (FLOAT_EPSILON, None))
+            kwargs['bounds']['y_stddev'] = kwargs['bounds'].get(
+                'y_stddev', (FLOAT_EPSILON, None))
+
```
Comment (astrofrog) [inline astropy/modeling/functional_models.py]: Any idea why simply specifying the bounds in Parameter doesn't work? Does it just get ignored? I think we should get to the bottom of this and either fix it or add a comment to explain why we can't do this.

Comment (pllim) [inline astropy/modeling/functional_models.py]: I think because of `Gaussian2D` re-defining the `__init__`? I am just guessing here. Convolution tests failed horribly when I put `bounds` in `Parameter`, to the point the tests just hang until I hit CTRL+C. Maybe @nden can explain.

Comment (saimn) [inline astropy/modeling/functional_models.py]: A minor stylistic note, I think this would be more readable:
```python
kwargs.setdefault('bounds', {})
kwargs['bounds'].setdefault('x_stddev', (FLOAT_EPSILON, None))
kwargs['bounds'].setdefault('y_stddev', (FLOAT_EPSILON, None))
```
Or you can use the 2 last lines in your `else` clause if you prefer.

**Comment by pllim** [comment]
Any chance this can make it into v2.0? Please re-review.

p.s. I am guessing I'll have to rebase again when #6183 is merged; so I'll take care of the change log weirdness after approval.

File: CHANGES.rst [hunk 7]
```diff
@@ -395,6 +395,19 @@ Bug Fixes
...
+
+- ``astropy.io.votable``
+
+- ``astropy.modeling``
+
+  - Gaussian models now impose positive ``stddev`` in fitting. [#6019]
+
+- ``astropy.nddata``
+
+- ``astropy.stats``
```
Comment (astrofrog) [inline CHANGES.rst]: This PR needs to be rebased and this entry added to 2.0 (I don't think there will be a 1.3.4)

File: astropy/modeling/functional_models.py [hunk 8]
```diff
@@ -365,6 +370,13 @@ def __init__(self, amplitude=amplitude.default, x_mean=x_mean.default,
                 y_vec = eig_vecs[:, 0]
                 theta = np.arctan2(y_vec[1], y_vec[0])
 
+        # Ensure stddev makes sense if its bounds are not explicitly set.
+        # stddev must be non-zero and positive.
+        # Setting this in Parameter above causes convolution tests to hang.
```
Comment (astrofrog) [inline astropy/modeling/functional_models.py]: I would add a TODO or FIXME here - this is clearly something that should be fixed in future.

File: astropy/modeling/tests/test_constraints.py [hunk 9]
```diff
@@ -23,6 +23,9 @@
 except ImportError:
     HAS_SCIPY = False
 
+# Module level random state for reproducible results.
+rsn = RandomState(1234567890)
```
Comment (astrofrog) [inline astropy/modeling/tests/test_constraints.py]: Module level won't work if people choose to run a single test, I think we need to set the random state inside each test that generates random numbers

File: astropy/stats/info_theory.py (hunk 2 above)
Comment (mirca) [inline astropy/stats/info_theory.py]: Hey @pllim, sorry, I didn't notice you mentioned me here.
Hmm, it's surprising that only this value changed...
But this shouldn't hold this PR. :)

Comment (pllim) [inline astropy/stats/info_theory.py]: Thanks! Well, that is the only *tested* value that changed in `stats`, yes. Those with "doc skip" are not tested.

File: astropy/stats/info_theory.py [hunk 10]
```diff
@@ -379,7 +379,7 @@ def akaike_info_criterion_lsq(ssr, n_params, n_samples):
...
     >>> akaike_info_criterion_lsq(ssr_g2, 6, x.shape[0]) # doctest: +FLOAT_CMP
     -662.83834510232043
     >>> akaike_info_criterion_lsq(ssr_g1, 3, x.shape[0]) # doctest: +FLOAT_CMP
```
Comment (mirca) [inline astropy/stats/info_theory.py]: @pllim I was susprised those two values bellow didn't change.

Comment (pllim) [inline astropy/stats/info_theory.py]: @mirca , for some reason, `g3_fit` had negative `mean_1` without this PR and positive with this PR. Do you have any objection to the new results?

Without this PR:
```python
In [2]: g3_fit
Out[2]: <CompoundModel1(amplitude_0=0.1755099144937883, mean_0=0.2418419988277658, stddev_0=0.28246312142011093, amplitude_1=0.14903703265951587, mean_1=-3.057112188121801, stddev_1=0.30735643205583174, amplitude_2=2.451435082571759, mean_2=0.503285183744267, stddev_2=0.09909043268097108)>

In [3]: g2_fit
Out[3]: <CompoundModel2(amplitude_0=0.1753819504769808, mean_0=0.24045925337616975, stddev_0=0.28126103361662386, amplitude_1=2.4523765774919726, mean_1=0.5032999106348427, stddev_1=0.09912276546074991)>

In [4]: g1_fit
Out[4]: <Gaussian1D(amplitude=2.5329026083160606, mean=0.49935793058448075, stddev=0.10559425938501048)>
```

With this PR:
```python
In [2]: g3_fit
Out[2]: <CompoundModel1(amplitude_0=0.5965686483533806, mean_0=0.5093102217176415, stddev_0=0.5549516598914004, amplitude_1=-0.48405948352535877, mean_1=1.0170405002235874, stddev_1=0.9359042247635705, amplitude_2=2.3912809994262276, mean_2=0.5021418528282614, stddev_2=0.09698667234256535)>

In [3]: g2_fit
Out[3]: <CompoundModel2(amplitude_0=0.1753819504769808, mean_0=0.24045925337616975, stddev_0=0.28126103361662386, amplitude_1=2.4523765774919726, mean_1=0.5032999106348427, stddev_1=0.09912276546074991)>

In [4]: g1_fit
Out[4]: <Gaussian1D(amplitude=2.5329026083160606, mean=0.49935793058448075, stddev=0.10559425938501048)>
```

**Comment by pllim** [comment]
@astrofrog , I applied your comments. Travis tests are running on my own fork (link given above in my original post). Not sure why Circle CI failed.

**Comment by bsipocz** [comment]
@pllim - circle is restarted, I didn't see an obvious reason why it failed the first time.

**Comment by pllim** [comment]
Thanks, @bsipocz . Circle CI passed. Travis on my fork also passed.

**Comment by bsipocz** [comment]
@pllim - Do you still waiting for one more review, or shall we merge?

**Comment by pllim** [comment]
AFAIK I addressed all the comments and I don't see any remaining objection. Do you want to push the big green button? :smile:

**Comment by bsipocz** [comment]
@pllim - With pleasure!
